tools/src/bin/run-fap.rs

```diff
@@ -0,0 +1,111 @@
+use std::{fmt, io, path::PathBuf, thread, time::Duration};
+
+use clap::Parser;
+use flipperzero_tools::{serial, storage};
+use rand::{thread_rng, Rng};
+
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+struct Cli {
+    /// Serial port (e.g. `COM3` on Windows or `/dev/ttyUSB0` on Linux)
+    #[arg(short, long)]
+    port: Option<String>,
+
+    /// Path to the FAP binary to run.
+    fap: PathBuf,
+
+    /// Arguments to provide to the FAP binary.
+    ///
+    /// Ignored until https://github.com/flipperdevices/flipperzero-firmware/issues/2505 is resolved.
+    args: Vec<String>,
+}
+
+enum Error {
+    FlipperZeroNotFound,
+    FailedToOpenSerialPort(serialport::Error),
+    FailedToStartSerialInterface(io::Error),
+    FailedToUploadFap(io::Error),
+    MkdirFailed(storage::FlipperPath, io::Error),
+    RemoveFailed(storage::FlipperPath, io::Error),
+    Io(io::Error),
+}
+
+impl fmt::Debug for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::FlipperZeroNotFound => write!(f, "unable to find Flipper Zero"),
+            Error::FailedToOpenSerialPort(e) => write!(f, "unable to open serial port: {}", e),
+            Error::FailedToStartSerialInterface(e) => {
+                write!(f, "unable to start serial interface: {}", e)
+            }
+            Error::FailedToUploadFap(e) => write!(f, "unable to upload FAP: {}", e),
+            Error::MkdirFailed(path, e) => write!(f, "unable to make directory '{}': {}", path, e),
+            Error::RemoveFailed(path, e) => write!(f, "unable to remove '{}': {}", path, e),
+            Error::Io(e) => e.fmt(f),
+        }
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(value: io::Error) -> Self {
+        Error::Io(value)
+    }
+}
+
+fn wait_for_idle(serial_cli: &mut serial::SerialCli) -> io::Result<()> {
+    loop {
+        serial_cli.send_and_wait_eol("loader info")?;
+        if serial_cli
+            .consume_response()?
+            .contains("No application is running")
+        {
+            break Ok(());
+        }
+        thread::sleep(Duration::from_millis(200));
+    }
+}
+
+fn main() -> Result<(), Error> {
+    let cli = Cli::parse();
+
+    let port_info =
+        serial::find_flipperzero(cli.port.as_deref()).ok_or(Error::FlipperZeroNotFound)?;
+    let port = serialport::new(&port_info.port_name, serial::BAUD_115200)
+        .timeout(Duration::from_secs(30))
+        .open()
+        .map_err(Error::FailedToOpenSerialPort)?;
+    let mut store = storage::FlipperStorage::new(port);
+    store.start().map_err(Error::FailedToStartSerialInterface)?;
+
+    // Upload the FAP to a temporary directory.
+    let dest_dir =
+        storage::FlipperPath::from(format!("/ext/tmp-{:08x}", thread_rng().gen::<u32>()));
```

Comment:
I haven't yet figured out if any of the Flipper Zero SDKs let a running external FAP learn where it was loaded from. Without that knowledge, it's gonna be tricky to reliably redirect stdout to a file that both won't collide with an existing file on the Flipper Zero, and that this tool can find.

Reply:
I ended up just using a common fixed path of `/ext/flipperzero-rs-stdout`, on the assumption that external FAPs cannot run concurrently, so only one app at a time will ever be running, and as long as we immediately read the file and then remove it we're probably fine. This will do until the Flipper Zero SDK supports RPC usage.

Reply:
You may try addressing the issues in Flipperzero [Telegram group](https://t.me/flipperzero), i.e. `Development (SW&HW)` channel.

Reply:
https://github.com/flipperdevices/flipperzero-firmware/issues/2505#issuecomment-1474184238 indicated that an improved host-device communication protocol is planned for external FAPs, so that is likely what will be used instead when available.